monitor/src/main.rs

```diff
@@ -0,0 +1,289 @@
+async fn get_blocks_from(
+    start: DateTime<Utc>,
+    web3: &Web3,
+) -> impl Stream<Item = anyhow::Result<Block<Transaction>>> + '_ {
+    let eth = web3.eth();
+    let mut block_number = BlockNumber::Latest;
+    let mut blocks = Vec::new();
+    let filter = EthFilter::new(web3.transport())
+        .create_blocks_filter()
+        .await
+        .expect("unable to register event listener for new blocks");
+    let new_blocks = filter.stream(BLOCK_POLL_DURATION);
+    loop {
+        let block = eth.block_with_txs(BlockId::Number(block_number)).await;
+        if let Err(e) = block {
+            tracing::warn!("node error when fetching the block: {:?}", e);
+            tokio::time::delay_for(NODE_RESPITE).await;
+            continue;
+        };
+        let block = block.unwrap();
+        if block.is_none() {
+            tracing::warn!("no block found for block number {:?}", block_number);
+            tokio::time::delay_for(NODE_RESPITE).await;
+            continue;
+        };
+        let block = block.unwrap();
+
+        tracing::info!(
+            "retrieving past block {:?} at timestamp {:?}, hash {:?}",
+            block.number,
+            block.timestamp,
+            block.hash
+        );
+
+        if Utc.timestamp(block.timestamp.as_u64() as i64, 0) < start {
+            tracing::info!(
+                "finished retrieving past blocks, first block outside of time range is block {:?}",
+                block_number,
+            );
+            break;
+        }
+
+        block_number = BlockNumber::from(
+            block
+                .number
+                .expect("block fetched by block number should have a number")
+                - 1,
+        );
+        blocks.push(Ok(block));
+    }
+
+    stream::iter(blocks.into_iter().rev()).chain(new_blocks.then(move |block_hash| {
+        let eth = web3.eth();
+        async move {
+            if let Err(e) = block_hash {
+                // Note: this can happen if the node unregisters the block
+                // filter. In this case, the new block filter might have to be
+                // regenerated. This is not a panic to see
+                tracing::error!("block filter did not return a valid block: {}", e);
+                bail!("invalid block filter");
+            }
+            let response = eth.block_with_txs(BlockId::Hash(block_hash?)).await?;
+            if response.is_none() {
+                tracing::warn!("failed to retrieve a block returned by the new block filter");
+                bail!("invalid block from block filter")
+            }
+            Ok(response.unwrap())
+        }
+    }))
+}
+
+struct Settlement {
+    time: DateTime<Utc>,
+    success: bool,
+}
+
+struct Count {
+    total: usize,
+    failures: usize,
+}
+
+struct SettlementFailureCounter {
+    // Note: settlements are not expected to be sorted by time because of uncle
+    // blocks.
+    settlements: Vec<Settlement>,
```

Comment:
If this makes other logic easier you could enforce the invariant that the vector is always sorted. To uphold the invariant you can use the binary_search functions https://doc.rust-lang.org/std/primitive.slice.html#method.binary_search .

Reply:
Having it sorted is nice because `settlements` could be a `VecDeque` and pruning could be done by popping from the left until all old transactions are removed, instead of traversing each time. In terms of readability I think it's better how it is now, in terms of efficiency I don't know, but anyway the numbers are very small, so I think it might be ok to leave it like this.

Reply:
Agree, I wouldn't worry about efficiency. Just about code readability.